### PR TITLE
Move GSamplingProfiler to CaptureData

### PR DESCRIPTION
**Pull request by florian-kuebler**
This moves the global Capture::GSamplingProfiler to CaptureData.
By being tight to the CaptureData (of the previous capture), lifetime
is more explicit.

Follow Up:
1. Make the shared_ptr in CaptureData a value type.
2. Move CaptureData to DataManager.
3. Make SamplingProfiler stateless (requires updates of SamplingReport).

Test: Compile, Take Capture, Load Symbols (before and after). Check all
 views. Load a Capture.
Bug: http://b/163020637

### CODE REVIEW

File: OrbitCore/CaptureData.h [hunk 1]
```diff
@@ -23,7 +24,8 @@ class CaptureData {
         process_{std::move(process)},
         selected_functions_{std::move(selected_functions)},
         callstack_data_(std::make_unique<CallstackData>()),
-        selection_callstack_data_(std::make_unique<CallstackData>()) {
+        selection_callstack_data_(std::make_unique<CallstackData>()),
+        sampling_profiler_{std::make_shared<SamplingProfiler>(process_)} {
```
Comment (dimitry-) [inline OrbitCore/CaptureData.h:28]: Is there a reason why this is not in DataManager? It seems like this should live separately, even though it uses the capture data and is tied to capture state (it uses capture process not selected process).

Comment (florian-kuebler) [reply OrbitCore/CaptureData.h:28]: The idea was that the lifetime is completely bound by the one of capture data (at least in theory).

Comment (dimitry-) [reply OrbitCore/CaptureData.h:28]: On the other hand we do not want to serialize it. And the lifetime is actually a bit different - the sampling profiler data can only be calculated once capture is done.

Comment (dimitry-) [reply OrbitCore/CaptureData.h:28]: discussed offline - decided to keep it here because we want it to be freed when capture data is deallocated.

File: OrbitCore/CaptureData.h [hunk 2]
```diff
@@ -125,6 +129,10 @@ class CaptureData {
 
   [[nodiscard]] const std::shared_ptr<Process>& process() const { return process_; }
 
+  [[nodiscard]] const std::shared_ptr<SamplingProfiler>& sampling_profiler() const {
```
Comment (dimitry-) [inline OrbitCore/CaptureData.h:132]: should this be a const object? std::shared_ptr<const SamplingProfiler>?

Comment (dimitry-) [reply OrbitCore/CaptureData.h:132]: or have 2 functions - one that returns const pointer/const reference another (non-const) which returns shared_ptr.

**Review by dimitry- — APPROVED** [review]
otherwise lgtm

File: OrbitCore/CaptureData.h (hunk 2 above)
Comment (florian-kuebler) [reply OrbitCore/CaptureData.h:132]: I added a second getter returning a const ref SamplingProfiler

File: OrbitGl/App.cpp [hunk 3]
```diff
@@ -612,12 +610,9 @@ void OrbitApp::ClearCapture() {
   set_selected_thread_id(-1);
   SelectTextBox(nullptr);
 
-  // Trigger deallocation of previous sampling related data.
-  // TODO(kuebler): Investigate why selected process needs to be passed here.
-  auto empty_sampling_profiler = std::make_shared<SamplingProfiler>(GetSelectedProcess());
-  AddSamplingReport(empty_sampling_profiler, nullptr);
```
Comment (IrinaShkviro) [inline OrbitGl/App.cpp:618]: Why it was removed?

File: OrbitCore/CaptureData.h (hunk 2 above)
Comment (IrinaShkviro) [reply OrbitCore/CaptureData.h:132]: I think that if you're using this function to update profiler it shouldn't be `const`. Also I would prefer to use 
`[[nodiscard]] SamplingProfiler* sampling_profiler()` or 
`[[nodiscard]] std::shared_ptr<SamplingProfiler> sampling_profiler()` if you need shared_ptr.

File: OrbitGl/App.cpp (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGl/App.cpp:618]: Sorry, this was unintended

File: OrbitCore/CaptureData.h (hunk 2 above)
Comment (florian-kuebler) [reply OrbitCore/CaptureData.h:132]: I'll make it a non-const ref, as it is also stored into the sampling report (and I want avoid storing raw pointer here).
In the future, this will be a value type and the second getter will go away.

**Comment by florian-kuebler** [comment]
@kisenshi Added you as a reviewer FYI.

File: OrbitCore/CaptureData.h (hunk 2 above)
Comment (florian-kuebler) [reply OrbitCore/CaptureData.h]: actually made it returning the shared_ptr as value.

File: OrbitGl/App.cpp (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGl/App.cpp:618]: For the record, I now moved the initialization of capture data to OnCaptureStarted and also clear capture data on ClearCapture, so that this is again safe and correct here.

File: OrbitClientGgp/ClientGgp.cpp [hunk 4]
```diff
@@ -65,8 +65,6 @@ bool ClientGgp::RequestStartCapture(ThreadPool* thread_pool) {
   // TODO(kuebler): right now selected_functions is only an empty placeholder,
   //  it needs to be filled separately in each client and then passed.
   absl::flat_hash_map<uint64_t, orbit_client_protos::FunctionInfo> selected_functions;
-  Capture::capture_data_ =
```
Comment (kisenshi) [inline OrbitClientGgp/ClientGgp.cpp:68]: Why is this removed? I still need Capture and capture_data for processing the data of the capture (which is not done yet but it is the next step)

File: OrbitCore/CaptureData.h [hunk 5]
```diff
@@ -125,6 +129,12 @@ class CaptureData {
 
   [[nodiscard]] const std::shared_ptr<Process>& process() const { return process_; }
 
+  [[nodiscard]] std::shared_ptr<SamplingProfiler> sampling_profiler() const {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/CaptureData.h]: Done

File: OrbitCaptureClient/CaptureEventProcessorProcessEventsFuzzer.cpp [hunk 6]
```diff
@@ -21,7 +21,8 @@ using orbit_client_protos::TimerInfo;
 
 class MyCaptureListener : public CaptureListener {
  private:
-  void OnCaptureStarted() override {}
+  void OnCaptureStarted(
+      int32_t, const absl::flat_hash_map<uint64_t, orbit_client_protos::FunctionInfo>&) override {}
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCaptureClient/CaptureEventProcessorProcessEventsFuzzer.cpp]: Thanks. Yes added the comments.

**Review by kisenshi — APPROVED** [review]
The changes in the ggp client look good to me, thanks!

**Review by IrinaShkviro — APPROVED** [review]
Lgtm, thanks for the change!
